design/kata-api-design.md

```diff
@@ -0,0 +1,143 @@
+# Kata API Design
+To fulfill the [kata design requirements](kata-design-requirements.md), and based on the disscusion on [Virtcontainers API extentions](https://docs.google.com/presentation/d/1dbGrD1h9cpuqAPooiEgtiwWDGCYhVPdatq7owsKHDEQ), the kata runtime library features following APIs:
+-  Sandbox based top API
+-  Storage and network hotplug API
+-  Plugin frameworks for external proprietary Kata runtime extensions
+-  Built-in shim and proxy types and capabilities
+
+## Sandbox Based API
+### Sandbox Management API
+-  CreateSandbox(ID, SandboxConfig)
+
+Create, start a sandbox and return the sandbox structure.
+-  ConnectSandbox(ID)
+
+Connect to an existing sandbox and return the sandbox structure.
+-  ListSandbox()
+
+List all existing standboxes with status.
+
+### Sandbox Operation API
+-  sandbox.Release()
```

Comment:
What is the implication of calling this API? Does the sandbox continue to exist?

Reply:
Yes. The sandbox can continue running and CRI shims can reconnect to it with the `ConnectSandbox()` API.

Reply:
Then you don't need a release, you can just use a `FetchPod()` when it has been restarted.

Reply:
So we never need graceful shutdown? hmmm, it looks reasonable at first glance since all APIs are stateless and the kata agent does not quit on connection errors. @laijs WDYT?

Reply:
sandbox.Release() is for the graceful shutdown in case there are some long-living goroutines or other resources needed to be cleanup attached to the sandbox. The process should be able to release to a sandbox and attach it again. The first in-memory sandbox object should not have any leftover that interferes the later one.

runv's sandbox has several goroutines attached. I'm not sure about cc-runtime's.